### ISSUE TITLE
External packages cannot access dependencies in spec

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#9149]
I'm trying to register `py-numpy` as an external package, but when I try to install it, it crashes saying that `python` is not in the spec.

### Steps to reproduce the issue

1. Register `py-numpy` as an external package
2. Install `py-numpy`

### Error Message

```console
$ spack --debug install py-numpy
==> py-numpy@1.14.2 : has external module in bwpy/1.2.4
==> py-numpy@1.14.2 : is actually installed in /opt/bwpy
==> py-numpy@1.14.2 : generating module file
Traceback (most recent call last):
  File "/u/sciteam/stewart1/spack/bin/spack", line 54, in <module>
    sys.exit(spack.main.main())
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/main.py", line 653, in main
    return _invoke_command(command, parser, args, unknown)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/main.py", line 432, in _invoke_command
    return_val = command(parser, args)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/cmd/install.py", line 273, in install
    install_spec(args, kwargs, spec)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/cmd/install.py", line 154, in install_spec
    spec.package.do_install(**kwargs)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/package.py", line 1380, in do_install
    return self._process_external_package(explicit)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/package.py", line 1302, in _process_external_package
    spack.hooks.post_install(self.spec)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/hooks/__init__.py", line 73, in __call__
    hook(*args, **kwargs)
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/hooks/module_file_generation.py", line 48, in <lambda>
    post_install = lambda spec: _for_each_enabled(spec, 'write')
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/hooks/module_file_generation.py", line 41, in _for_each_enabled
    getattr(generator, method_name)()
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/modules/common.py", line 698, in write
    context = self.context.to_dict()
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/tengine.py", line 86, in to_dict
    d = [(name, getattr(self, name)) for name in self.context_properties]
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/modules/common.py", line 553, in environment_modifications
    self.spec.package.setup_environment(_, env)
  File "/mnt/a/u/sciteam/stewart1/spack/var/spack/repos/builtin/packages/py-numpy/package.py", line 159, in setup_environment
    
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/spec.py", line 2817, in __getitem__
    raise KeyError("No spec with name %s in %s" % (name, self))
KeyError: 'No spec with name python in py-numpy@1.14.2%cce@8.4.6+blas+lapack arch=cray-cnl5-interlagos'
```
The problem is that the `py-numpy` package contains the following `setup_environment`:
```python
    def setup_environment(self, spack_env, run_env):
        python_version = self.spec['python'].version.up_to(2)

        include_path = join_path(
            self.prefix.lib,
            'python{0}'.format(python_version),
            'site-packages',
            'numpy/core/include')

        run_env.prepend_path('CPATH', include_path)
```
When creating a module file, Spack queries the `python` version. But dependencies don't seem to make it into the `spec` for external packages, causing the `KeyError`.

@alalazo

**Comment by scheibelp** [comment]
This may be tricky, I can start looking at it on 9/4. Generally external packages have all dependencies truncated.

Do you also have an externally-installed Python? If so, is it also listed in your `packages.yaml`?

**Comment by scheibelp** [comment]
> a simple solution to this would be to skip module file generation for external packages.

That would work here but is arguably roundabout. If we can agree on the conditions when to avoid modules, and those apply to your case, then I'm fine with that.

I think that in general @alalazo wants to generate modules for external packages. If your external `py-numpy` has an associated module file, and if we can agree that externals based on modules shouldn't generate Spack modules, that would work. If you don't have a custom module but would be willing to generate one, that would also work.

If that isn't satisfactory:

I'm considering whether Spack should try to assign external dependencies to external dependents: while skipping module generation would work here, it would not work for any external dependent of Python which does something similar in `setup_dependent_environment`. For example if `X->Y->Python`, and Y is external, then a similar error would occur when the build for `X` invokes `setup_dependent_environment` for `Y`.

A simpler option from the perspective of core is to decide that by convention packages should guard against the possibility that their dependencies don't exist.

Between those two options I favor the first: the Python Spack package contains logic for handling external Python installs which may benefit external dependents. That may be difficult to handle without additional hints in `packages.yaml` though.

**Comment by scheibelp** [comment]
> If your external py-numpy has an associated module file, and if we can agree that externals based on modules shouldn't generate Spack modules, that would work.

This is now implemented in https://github.com/spack/spack/pull/7759

**Comment by goxberry** [comment]
I've also run into this bug; however, my external `py-numpy` is not associated with a module file, but rather it's installed using a "system" package manager (macOS, Homebrew). Is adding dependency information from specs into modules on the future development roadmap? I recognize the team is busy; I ask so I can make a note in my `packages.yaml` about whether this issue may be resolved in the future, so my future self doesn't forget and stubbornly try to add `py-numpy` as an external again.

**Comment by goxberry** [comment]
I think I've run into this bug again in trying to kludge around #10493 to build `mfem`. The `mfem` spack package has a line https://github.com/spack/spack/blob/514cf0c1f4d0c7c33bb138f061507638891f1533/var/spack/repos/builtin/packages/mfem/package.py#L276 that queries the `hypre` spec for its `blas` and `lapack` dependencies. If I add `hypre@2.15.1+mpi` as an external package, then spack throws an error:

```console
==> Error: KeyError: 'No spec with name lapack in hypre@2.15.1%intel@18.0.1+debug~int64+internal-superlu+mpi+shared arch=linux-rhel7-x86_64'
```

even though the external package in `${SPACK_ROOT}/etc/spack/packages.yaml` has the snippet

```yaml
packages:
## ... snipped lines ... ##
  hypre:
    paths:
      hypre@2.15.1+debug~int64+internal-sperlu+mpi+shared%intel@18.0.1 ^netlib-lapack@3.4.2 ^mvapich2@2.2%intel@18.0.1: /collab/usr/gapps/librom/hypre-2.15.1
## ... snipped lines ... ##
  mvapich2:
    paths:
      mvapich2@2.2%intel@18.0.1: /usr/tce/packages/mvapich2/mvapich2-2.2-intel-18.0.1
## ... snipped lines ... ##
  netlib-lapack:
    paths:
      netlib-lapack@3.4.2: /usr/
## ... more snipped lines ... ##
```

because external package dependencies are ignored.

**Comment by kshitij-v-mehta** [comment]
Tagging onto this issue as I am facing the same on the Summit machine @ ORNL:
My packages.yaml has the Python module as an external package, along with mpi4py and py-numpy.
```
 # snipped #
    python:
        modules:
            python@3.7: python/3.7.0-anaconda3-5.3.0
        buildable: False
    py-mpi4py:
        paths:
            py-mpi4py@3.0.0: /sw/summit/python/3.7/anaconda3/5.3.0
        buildable: False
    py-numpy:
        paths:
            py-numpy@1.14.5: /sw/summit/python/3.7/anaconda3/5.3.0
        buildable: False
# snipped #
```

Installing adios2 gives
`==> Error: 'No spec with name python in py-numpy@1.14.5%gcc@8.1.1+blas+lapack arch=linux-rhel7-ppc64le'`

**Comment by adamjstewart** [comment]
See #10881 and #12050 for workarounds to avoid this problem until we come up with a better solution.

**Comment by adamjstewart** [comment]
@scheibelp this bug is cropping up in quite a few places (see list of all the issues/PRs that have referenced this issue). Is this issue still on the radar?

**Comment by scheibelp** [comment]
Thanks for the ping, I've added it to our Wednesday meeting agenda at https://github.com/spack/spack/wiki/Telcon:-2022-01-26

https://github.com/spack/spack/pull/24894 might partially address this (although it requires that the user manually specify the dag themselves).

It's possible we could also infer connections between dependencies, although that would likely require more extensive work on making packages detectable, and iterating on https://github.com/spack/spack/pull/28005 (which adds preliminary support for detecting external libraries but I think more work would be needed for it to address this particular case).

**Comment by alalazo** [comment]
@adamjstewart I wouldn't regard this as a bug, but as a feature request :slightly_smiling_face: The current model for externals is that we truncate all the dependencies (and don't account for conflicts) and we get information from `packages.yaml`. What would be ideal is that:
1. We could detect externals automatically via `spack external find`
2. Each detection has means to detect the entire DAG
3. Externals gets registered directly in the DB, rather than in `packages.yaml`

There are several open questions though with this, like how to best deal with sandboxing of externals (you might want to seek for a `foo` external which lives in the same directory as many other externals you don't want to use) or how to recurse  on dependencies for software that is already installed.

The last issue requires us to be able to reconstruct a DAG from artifacts that are already on disk, rather than installing artifacts for a DAG that we solved for ourselves.

**Comment by adamjstewart** [comment]
Presumably closed by #33777
